docs: :memo: start of guide to creating a package (#557)
seedcase-project/seedcase-sprout: docs/guide/packages.qmd

---
title: "Creating and Managing Data Packages"
---

{{< include /docs/includes/_wip.qmd >}}

At the core of Sprout is the Data Package, which is a standardized way
to structuring and sharing data. This guide will show you how to create
and manage data packages using Sprout.

<!-- TODO: this callout might need to be its own include, since other docs will use it too -->

::: callout-important
For both the Python library and the CLI, Sprout assumes you have full
control over the folders and files of the system, or at least your
user's home directory. This includes being given space on a server that
mostly has access through a Terminal, where you have control over the
directories you can write to.

An easy example of this is if you install Sprout on your own computer
because you want to create some data packages for studies you are
running. Another example would be if your research group has storage
space on a server that you need to use a Terminal and SSH to access.
:::

## Creating a Data Package

:::: panel-tabset
### Python

> The first thing you'll need to decide is where you want to store all

Reviewer: ```suggestion
The first thing you'll need to decide is where you want to store
```

your data packages. By default, Sprout will create it in
`~/.sprout/packages/` on Linux (see [Outputs](design/outputs.qmd) for
operating system specific locations), but you can change this by setting
the `SPROUT_ROOT` environment variable. For instance, maybe you want the
location to be `~/Documents/data-packages/`. You can set this in your
Python script like so:

``` python
import sprout.core as sp
import os

os.environ["SPROUT_ROOT"] = "~/Documents/data-packages/"
```

> Afterwards, you would create the structure for your first data package

Reviewer: ```suggestion
Afterwards, you will create the structure for your first data package
```

like so:

``` python
sp.create_package_structure(path=sp.path_package())
```

```         
[PosixPath('~/Documents/data-packages/1/datapackage.json'),
 PosixPath('~/Documents/data-packages/1/README.md'),
 PosixPath('~/Documents/data-packages/1/resources')]
```

Then to fill in the details in the `datapackage.json` file, you can use
the `create_package_properties()` function:

``` python
package_properties = sp.create_package_properties(
  path=sp.path_package_properties(package_id=1),
  properties=sp.properties_json(
    title="Diabetes and Hypertension Study", 
    description="Data from the 2021 study on diabetes and hypertension", 

>     contributors=["Jamie Jones", "Jimmy James"],

Reviewer: ```suggestion
    contributors=[{
  "title": "Jamie Jones",
  "email": "jamie_jones@example.com",
  "roles": ["creator"]}, 
  {"title": "Jimmy James",
  "email": "jimmy_james@example.com",
  "roles": ["contact"]
  }] 
```
To follow something like Data Package show in their docs [here](https://datapackage.org/standard/data-package/#contributors)

    licenses=["ODC-BY-1.0"]
  )
)

# TODO: add an example output of the above.

> # Save the package properties back to the file

Reviewer: ```suggestion
# Save the package properties to the datapackage.json file
```

sp.write_package_properties(
  json_object=package_properties,
  path=sp.path_package_properties(package_id=1)
)
```

If you need help with filling in the right properties, use
`view_package_properties_template()` to see the available fields to fill
in.

### CLI

The CLI is a bit more straightforward as long as you are comfortable
using the Terminal. You can set the `SPROUT_ROOT` environment variable
to change the location of the data packages. For instance, maybe you
want the location to be `~/Documents/data-packages/`. You can set this
in your terminal like so:

``` bash
export SPROUT_ROOT=~/Documents/data-packages/
```

Then creating a new package would be as simple as:

``` bash
sprout package create 
```

This will prompt you for some required fields you need to fill in, like
the title and description of the data package. If you want to skip the
prompt, you can provide the information directly in the command:

```         

> sprout package create --title "Diabetes and Hypertension Study" \

Reviewer: Do we want `contributors` to be a required field? So there's always someone's name (preferably the creator of the package) in `datapackage.json`?

Author: I think we can decide on that when we actually get the full spec and incorporate it into the functions. I think we should download the full spec, show the full spec, and allow users to edit as needed. At least for now, make the required data package fields required by us as well, and decide/discuss as we go along for other fields.

Reviewer: Sounds like a good plan 👍 
About required fields: A Data Package is required to have the [resources](https://datapackage.org/standard/data-package/#resources) property. How will that work when creating a package? Will we create the Data Package with an empty resources property?

Author: That's right!

---
title: "Creating and managing Data Packages"
---

{{< include /docs/includes/_wip.qmd >}}

At the core of Sprout is the Data Package, which is a standardized way

> to structuring and sharing data. This guide will show you how to create

Reviewer 2: ```suggestion
of structuring and sharing data. This guide will show you how to create
```

---
title: "Creating and managing Data Packages"
---

{{< include /docs/includes/_wip.qmd >}}

At the core of Sprout is the Data Package, which is a standardized way
of structuring and sharing data. This guide will show you how to create
and manage data packages using Sprout.

<!-- TODO: this callout might need to be its own include, since other docs will use it too -->

::: callout-important
For both the Python library and the CLI, Sprout assumes you have full
control over the folders and files of the system, or at least your
user's home directory. This includes being given space on a server that
mostly has access through a Terminal, where you have control over the
directories you can write to.

An easy example of this is if you install Sprout on your own computer

> because you want to create some data packages for research studies you

Author: Hmm, good point. I've used the capitalization wrong here. I'll fix that later. It should be lower case unless referring (and linking) directly to the website and standard.

are running. Another example would be if your research group has storage
space on a server that you need to use a Terminal and SSH to be able to
access.
:::

## Creating a Data Package

:::: panel-tabset
### Python

The first thing you'll need to decide is where you want to store your
data packages. By default, Sprout will create it in
`~/.sprout/packages/` on Linux (see [Outputs](/docs/design/outputs.qmd)
for operating system specific locations), but you can change this by
setting the `SPROUT_ROOT` environment variable. For instance, maybe you
want the location to be `~/Documents/data-packages/`. You can set this
in your Python script like so:

``` python
import sprout.core as sp
import os

os.environ["SPROUT_ROOT"] = "~/Documents/data-packages/"
```

> Afterwards, you will create the structure for your first data package by

Reviewer: ```suggestion
Afterwards, you can create the structure for your first data package by
```

using:

``` python
sp.create_package_structure(path=sp.path_package())
```

```         
[PosixPath('~/Documents/data-packages/1/datapackage.json'),
 PosixPath('~/Documents/data-packages/1/README.md')]
```

> 

Reviewer: Add something to explain what just happened?
... will the `datapackage.json` be empty initially?
```suggestion

This creates the initial structure of your new package with the ID `1`. The output above shows that the folder of your Data Package `1` has been created. This folder consists of two files: `datapackage.json` and `README.md`. The `datapackage.json` file is empty initially, but it will contain the metadata, a.k.a. the `properties`, of your Data Package. `README.md` is a prettified, human-readable version of the content of the `datapackage.json`.
```

Then to fill in the details in the `datapackage.json` file, you can use
the `create_package_properties()` function:

``` python
package_properties = sp.create_package_properties(
  path=sp.path_package_properties(package_id=1),

>   properties=sp.properties_json(

Reviewer: I don't recall seeing the `properties_json()` function in the core functions? Shouldn't it be included there?

Author: Good point. I'll add it to the helper functions like `pretty_json()`.

    title="Diabetes and Hypertension Study", 
    description="Data from the 2021 study on diabetes and hypertension", 
    contributors=[{
        "title": "Jamie Jones",
        "email": "jamie_jones@example.com",
        "roles": ["creator"]
      }, 
      {
        "title": "Jimmy James",
        "email": "jimmy_james@example.com",
        "roles": ["contact"]
      }],
    licenses=["ODC-BY-1.0"]
  )
)

# TODO: add an example output of the above.

> # Save the package properties to the `datapackage.json` file
> sp.write_package_properties(
>   json_object=package_properties,
>   path=sp.path_package_properties(package_id=1)
> )
> ```

Reviewer: Argh, the suggestion is cut off bc of the use of ``` to do the code blocks. 
```suggestion
```
To save the package properties to the `datapackage.json` file, run:

```
sp.write_package_properties(
  json_object=package_properties,
  path=sp.path_package_properties(package_id=1)
)
```
```

Author: If you ever want to do something like that, add an extra backtick to the `suggestion` block. So it is wrapped with something like:

`````
````suggestion
````
`````

If you need help with filling in the right properties, use
`view_package_properties_template()` to see the available fields to fill
in.

You now have the basic starting point for adding data resources to your
data package.

### CLI

The CLI is a bit more straightforward as long as you are comfortable
using the Terminal. You can set the `SPROUT_ROOT` environment variable
to change the location of the data packages. For instance, maybe you
want the location to be `~/Documents/data-packages/`. You can set this
in your terminal like so:

``` bash
export SPROUT_ROOT=~/Documents/data-packages/
```

Then creating a new package would be as simple as:

``` bash
sprout package create 
```

This will prompt you for some required fields you need to fill in, like
the title and description of the data package. If you want to skip the
prompt, you can provide the information directly in the command:

``` bash
sprout package create \
  --title "Diabetes and Hypertension Study" \
  --description "Data from the 2021 study on diabetes and hypertension"
```

> 

Reviewer: Again: Might be nice with some context abotu what they have just created?
```suggestion

This creates the initial structure of your new package with the ID `1`. The output above shows that the folder of your Data Package `1` has been created. This folder consists of two files: `datapackage.json` and `README.md`. The `datapackage.json` file is empty initially, but it will contain the metadata, a.k.a. the `properties`, of your Data Package. `README.md` is a prettified, human-readable version of the content of the `datapackage.json`.

```

> You will now have a new data package in the location you specified:

Reviewer: ```suggestion
You now have the basic starting point for adding data resources to your data package in the location you specified:
```
